File: platform_src/ios/elastOS/Classes/MergeDBAdapter.swift
 /*
  * Copyright (c) 2020 Elastos Foundation
  *
  * Permission is hereby granted, free of charge, to any person obtaining a copy
  * of this software and associated documentation files (the "Software"), to deal
  * in the Software without restriction, including without limitation the rights
  * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  * copies of the Software, and to permit persons to whom the Software is
  * furnished to do so, subject to the following conditions:
  *
  * The above copyright notice and this permission notice shall be included in all
  * copies or substantial portions of the Software.
  *
  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */

import Foundation
import SQLite

class MergeDBAdapter {
    let baseDBAdapter: ManagerDBAdapter;
    var userDBAdapter: ManagerDBAdapter? = nil;
    
    init(_ dataPath: String) {
        baseDBAdapter = ManagerDBAdapter(dataPath);
    }
    
    func setUserDBAdapter(_ databasePath: String?) {
        if (databasePath != nil) {
            userDBAdapter = ManagerDBAdapter(databasePath!);
        }
        else {
            userDBAdapter = nil;
        }
    }

    func addAppInfo(_ info: AppInfo, _ isShare: Bool) throws {
        if (isShare || info.built_in || userDBAdapter == nil) {
            return try baseDBAdapter.addAppInfo(info);
        }
        else {
            return try userDBAdapter!.addAppInfo(info);
        }
    }

    func getAppInfo(_ id: String) throws -> AppInfo? {
        var info: AppInfo? = nil;
        if (userDBAdapter != nil) {
            info = try userDBAdapter!.getAppInfo(id);
        }
        if (info == nil) {
            info = try baseDBAdapter.getAppInfo(id);
            if (info != nil && userDBAdapter != nil) {
                try userDBAdapter!.getAppAuthInfo(info!);
            }
        }
        else {
            info!.share = false;
        }
        return info;
    }

    func getAppInfos() throws -> [AppInfo] {
        var list = [AppInfo]();

        var infos: [AppInfo] = [AppInfo]();
        if (userDBAdapter != nil) {
            infos = try userDBAdapter!.getAppInfos();
        }
        for info in infos {
            info.share = false;
            list.append(info);
        }

        let baseInfos = try baseDBAdapter.getAppInfos();
        for baseInfo in baseInfos {
            var needAdd = true;
            for info in infos {
                if (baseInfo.app_id == info.app_id) {
                    needAdd = false;
                    break;
                }
            }
            if (needAdd) {
                list.append(baseInfo);
                if (userDBAdapter != nil) {
                    try userDBAdapter!.getAppAuthInfo(baseInfo);
                }
            }
        }
        
        return list;
    }

    func getLauncherInfo() throws -> AppInfo? {
        return try baseDBAdapter.getLauncherInfo();
    }

    func changeBuiltInToNormal(_ appId: String) throws {
        return try baseDBAdapter.changeBuiltInToNormal(appId);
    }

    func updatePluginAuth(_ item: PluginAuth, _ auth: Int) throws {
        if (userDBAdapter != nil) {
            return try userDBAdapter!.updatePluginAuth(item, auth);
        }
    }

    func updateUrlAuth(_ tid: Int64, _ url: String, _ auth: Int) throws {
        if (userDBAdapter != nil) {
            return try userDBAdapter!.updateUrlAuth(tid, url, auth);
        }
    }

    func removeAppInfo(_ info: AppInfo, _ isShare: Bool) throws {
        if (userDBAdapter != nil && !isShare) {
            return try userDBAdapter!.removeAppInfo(info);
        }
        else {
            return try baseDBAdapter.removeAppInfo(info);
        }
    }

    func getIntentFilter(_ act: String) throws -> [String] {
        var list = [String]();

        var ids: [String] = [String]();
        if (userDBAdapter != nil) {
            ids = try userDBAdapter!.getIntentFilter(act);
        }
        for id in ids {
            list.append(id);
        }

        let baseIds = try baseDBAdapter.getIntentFilter(act);
        for baseId in baseIds {
            var needAdd = true;
            for id in ids {
                if (baseId == id) {
                    needAdd = false;
                    break;
                }
            }
            if (needAdd) {
                list.append(baseId);
            }
        }
        
        return list;
    }

    func setSetting(_ id: String, _ k: String, _ v: Any?) throws {
        if (userDBAdapter != nil) {
            return try userDBAdapter!.setSetting(id, k, v);
        }
        else {
            return try baseDBAdapter.setSetting(id, k, v);
        }
    }

    func getSetting(_ id: String, _ k: String) throws -> [String: Any]? {
        if (userDBAdapter != nil) {
            return try userDBAdapter!.getSetting(id, k);
        }
        else {
            return try baseDBAdapter.getSetting(id, k);
        }
    }

    func getSettings(_ id: String) throws -> [String: Any] {
        if (userDBAdapter != nil) {
            return try userDBAdapter!.getSettings(id);
        }
        else {
            return try baseDBAdapter.getSettings(id);
        }
    }

    func setPreference(_ k: String, _ v: Any?) throws {
        if (userDBAdapter != nil) {
            return try userDBAdapter!.setPreference(k, v);
        }
        else {
            return try baseDBAdapter.setPreference(k, v);
        }
    }

    func resetPreferences() throws {
        if (userDBAdapter != nil) {
            try userDBAdapter!.resetPreferences();
        }
        else {
            try baseDBAdapter.resetPreferences();
        }
    }

    func getPreference(_ k: String) throws -> [String: Any]? {
        if (userDBAdapter != nil) {
            return try userDBAdapter!.getPreference(k);
        }
        else {
            return try baseDBAdapter.getPreference(k);
        }
    }

    func getPreferences() throws -> [String: Any] {
        if (userDBAdapter != nil) {
            return try userDBAdapter!.getPreferences();
        }
        else {
            return try baseDBAdapter.getPreferences();
        }
    }
    
    func getApiAuth(_ appId: String, _ plugin: String, _ api: String) throws -> Int? {
        if (userDBAdapter != nil) {
            return try userDBAdapter!.getApiAuth(appId, plugin, api);
        }
        else {
            return try baseDBAdapter.getApiAuth(appId, plugin, api);
        }
    }
    
    func setApiAuth(_ appId: String, _ plugin: String, _ api: String, _ auth: Int?) throws {
        if (userDBAdapter != nil) {
            return try userDBAdapter!.setApiAuth(appId, plugin, api, auth);
        }
        else {
            return try baseDBAdapter.setApiAuth(appId, plugin, api, auth);
        }
    }
    
    func resetApiDenyAuth(_ appId: String)  {
        if (userDBAdapter != nil) {
            userDBAdapter!.resetApiDenyAuth(appId);
        }
        else {
            baseDBAdapter.resetApiDenyAuth(appId);
        }
    }
 }

File: platform_src/electron/main/MergeDBAdapter.ts
import { AppInfo } from "./AppInfo";
import { ManagerDBAdapter } from "./ManagerDBAdapter";
import { BrowserWindow } from 'electron';
import { IntentFilter } from './IntentFilter';
import { Setting } from './Setting';
import { Preference } from './Preference';

export class MergeDBAdapter {
    window: BrowserWindow = null;
    baseDBAdapter: ManagerDBAdapter = null;
    userDBAdapter: ManagerDBAdapter = null;

    constructor(window: BrowserWindow) {
        this.window = window;
        this.baseDBAdapter = new ManagerDBAdapter(window);
    }

    public static async newInstance(window: BrowserWindow): Promise<MergeDBAdapter> {
        let mergeDBAdapter = new MergeDBAdapter(window);
        await mergeDBAdapter.init();
        return mergeDBAdapter;
    }

    private async init() {
        this.baseDBAdapter = await ManagerDBAdapter.newInstance(this.window);
    }

    public async setUserDBAdapter(path: string) {
        if (path != null) {
            this.userDBAdapter = await ManagerDBAdapter.newInstance(this.window, path);
        }
        else {
            this.userDBAdapter = null;
        }
    }

    public async addAppInfo(info: AppInfo, isShare: boolean): Promise<boolean> {
        if (info != null) {
            if (isShare || info.built_in == 1 || this.userDBAdapter == null) {
                //console.log("Adding app info to BASE DB")
                return await this.baseDBAdapter.addAppInfo(info);
            }
            else {
                //console.log("Adding app info to USER DB")
                return await this.userDBAdapter.addAppInfo(info);
            }
        }
        else {
            return false;
        }
    }

    public async getAppInfo(id: string): Promise<AppInfo> {
        let info: AppInfo = null;
        if (this.userDBAdapter != null) {
            info = await this.userDBAdapter.getAppInfo(id);
        }
        if (info == null) {
            info = await this.baseDBAdapter.getAppInfo(id);
            if (info != null && this.userDBAdapter != null) {
                this.userDBAdapter.getAppAuthInfo(info);
            }
        }
        else {
            info.share = false;
        }
        return info;
    }

    public async getAppInfos(): Promise<AppInfo[]> {
        let list = new Array<AppInfo>();
        let infos = new Array<AppInfo>();
        if (this.userDBAdapter != null) {
            infos = await this.userDBAdapter.getAppInfos();
        }
        //console.log("getAppInfos after USER db:", infos.length);

        for (let info of infos) {
            info.share = false;
            list.push(info);
        }

        let baseInfos = await this.baseDBAdapter.getAppInfos();
        //console.log("getAppInfos after BASE db:", baseInfos.length);
        for (let baseInfo of baseInfos) {
            let needAdd = true;
            for (let info of infos) {
                if (baseInfo.app_id == info.app_id) {
                    needAdd = false;
                    break;
                }
            }
            if (needAdd) {
                list.push(baseInfo);
                if (this.userDBAdapter != null) {
                    await this.userDBAdapter.getAppAuthInfo(baseInfo);
                }
            }
        }
 
        return list;
    }

    public async getLauncherInfo(): Promise<AppInfo> {
        return this.baseDBAdapter.getLauncherInfo();
    }

    public async changeBuiltInToNormal(appId: string): Promise<number> {
        return this.baseDBAdapter.changeBuiltInToNormal(appId);
    }

    public async updatePluginAuth(tid: string, plugin: string, authority: number): Promise<number> {
        if (this.userDBAdapter != null) {
            return this.userDBAdapter.updatePluginAuth(tid, plugin, authority);
        }
        else {
            return 0;
        }
    }

    public async updateURLAuth(tid: string, url: string, authority: number): Promise<number> {
        if (this.userDBAdapter != null) {
            return this.userDBAdapter.updateURLAuth(tid, url, authority);
        }
        else {
            return 0;
        }
    }

    public async updateIntentAuth(tid: string, url: string, authority: number): Promise<number> {
        if (this.userDBAdapter != null) {
            return this.userDBAdapter.updateIntentAuth(tid, url, authority);
        }
        else {
            return 0;
        }
    }

    public removeAppInfo(info: AppInfo, isShare: boolean): Promise<number> {
        if (this.userDBAdapter != null && !isShare) {
            return this.userDBAdapter.removeAppInfo(info);
        }
        else {
            return this.baseDBAdapter.removeAppInfo(info);
        }
    }

    public async getIntentFilter(action: string): Promise<IntentFilter[]> {
        let list = new Array<IntentFilter>();
        let filters = null;
        if (this.userDBAdapter != null) {
            filters = await this.userDBAdapter.getIntentFilter(action);
        }

        let baseFilters = await this.baseDBAdapter.getIntentFilter(action);

        for (let filter of filters) {
            list.push(filter);
        }

        for (let baseFilter of baseFilters) {
            let needAdd = true;
            for (let filter of filters) {
                if (baseFilter.packageId == filter.packageId) {
                    needAdd = false;
                    break;
                }
            }
            if (needAdd) {
                list.push(baseFilter);
            }
        }

        return list;
    }

    public async setSetting(id: string, key: string, value: any): Promise<number> {
        if (this.userDBAdapter != null) {
            return this.userDBAdapter.setSetting(id, key, value);
        }
        else {
            return this.baseDBAdapter.setSetting(id, key, value);
        }
    }

    public async getSetting(id: string, key: string): Promise<Setting> {
        if (this.userDBAdapter != null) {
            return this.userDBAdapter.getSetting(id, key);
        }
        else {
            return this.baseDBAdapter.getSetting(id, key);
        }
    }

    public async getSettings(id: string): Promise<Setting[]> {
        if (this.userDBAdapter != null) {
            return this.userDBAdapter.getSettings(id);
        }
        else {
            return this.baseDBAdapter.getSettings(id);
        }
    }

    public async setPreference(key: string, value: any): Promise<number> {
        if (this.userDBAdapter != null) {
            return this.userDBAdapter.setPreference(key, value);
        }
        else {
            return this.baseDBAdapter.setPreference(key, value);
        }
    }

    public async resetPreferences() {
        if (this.userDBAdapter != null) {
            this.userDBAdapter.resetPreferences();
        }
        else {
            this.baseDBAdapter.resetPreferences();
        }
    }

    public async getPreference(key: string): Promise<any> {
        if (this.userDBAdapter != null) {
            return this.userDBAdapter.getPreference(key);
        }
        else {
            return this.baseDBAdapter.getPreference(key);
        }
    }

    public async getPreferences(): Promise<any> {
        if (this.userDBAdapter != null) {
            return this.userDBAdapter.getPreferences();
        }
        else {
            return this.baseDBAdapter.getPreferences();
        }
    }

    public async getApiAuth(appId: string, plugin: string, api: string): Promise<number> {
        if (this.userDBAdapter != null) {
            return this.userDBAdapter.getApiAuth(appId, plugin, api);
        }
        else {
            return this.baseDBAdapter.getApiAuth(appId, plugin, api);
        }
    }

    public async setApiAuth(appId: string, plugin: string, api: string, auth: number): Promise<number> {
        if (this.userDBAdapter != null) {
            return this.userDBAdapter.setApiAuth(appId, plugin, api, auth);
        }
        else {
            return this.baseDBAdapter.setApiAuth(appId, plugin, api, auth);
        }
    }

    public async resetApiDenyAuth(appId: string) {
        if (this.userDBAdapter != null) {
            this.userDBAdapter.resetApiDenyAuth(appId);
        }
        else {
            this.baseDBAdapter.resetApiDenyAuth(appId);
        }
    }

}
